### PR TITLE
diff mining with other changes

### PR DESCRIPTION
**Pull request by Butanium**
#46 but with all my changes on main and on greg branch

### CODE REVIEW

**Comment by Butanium** [comment]
# PR Review: `merged_juju_with_tests` -> `main`

**Date:** 2026-02-06
**Scope:** ~21,300 lines added, ~680 removed across 117 files (tests excluded from review)
**Reviewers:** 6 parallel code review agents covering different areas

---

## Executive Summary

This PR introduces the **diff_mining** method (top-K logit diff token analysis with NMF topic clustering), adds vLLM integration to the base `DiffingMethod` class, extends the agent/evaluation system, and includes extensive experiment scripts. The core architecture is sound, but there are several bugs and issues to address before merge.

**Verdict: Approve with requested changes** — 5 bugs to fix, several improvements recommended.

---

## Critical Issues (Must Fix)

### 1. `total_tokens` accumulation bug — doubles every iteration
**File:** `src/diffing/utils/agents/base_agent.py:314`
**Confidence: 95%**

```python
total_tokens += total_tokens  # BUG: doubles itself instead of adding new usage
```

This causes the `agent_total_tokens` stat to grow exponentially (2x after first call, 4x after second, etc.). Should be:
```python
total_tokens += int(usage.get("total_tokens", 0))
```

---

### 2. Missing `disable_compile` in `create_model_config()`
**File:** `src/diffing/utils/configs.py:142-168`
**Confidence: 100%**

`create_model_config()` doesn't extract `disable_compile` or `vllm_kwargs` from the DictConfig, even though:
- `ModelConfig` declares both fields
- `disable_compile` is set in all model YAML configs (e.g., Gemma models need `disable_compile=true`)

The field silently defaults to `False`, so models that require `disable_compile=true` will break at runtime. Fix: add both fields to `create_model_config()` and propagate to `finetuned_model_cfg`.

---

### 3. Silent TypeError catch hides real bugs in agent tools
**File:** `src/diffing/utils/agents/base_agent.py:426-440`
**Confidence: 95%**

```python
try:
    tool_output = tool_callable(**call_args)
except TypeError as e:
    error_msg = f"TOOL_PARAMETER_ERROR: ..."
    messages.append({"role": "user", "content": error_msg})
    continue
```

Violates fail-fast principle. If a tool has a genuine implementation bug (wrong parameter name, missing arg), this catch hides it and lets the agent waste turns retrying. The LLM argument validation should happen before the call, not by catching TypeError.

---

### 4. Memory risk: `.todense()` on potentially huge sparse matrix
**File:** `src/diffing/methods/diff_mining/token_ordering.py:503-510`
**Confidence: 95%**

The orthogonal NMF path converts a sparse COO matrix to dense before passing to `fit_nmf_orthogonal`:
```python
torch.tensor(scipy.sparse.coo_matrix(...).todense(), dtype=torch.float32)
```

For large vocab/dataset combinations (100k x 10k), this creates a ~4GB dense matrix. The non-orthogonal NMF path correctly keeps data sparse throughout. Fix: either pass sparse directly or add a memory check/warning.

---

### 5. Documentation default mismatch
**File:** `CLAUDE.md:8` vs `configs/config.yaml:4`
**Confidence: 100%**

CLAUDE.md says the default method is `activation_difference_lens`, but `config.yaml` sets it to `diff_mining`. Users following docs will have wrong expectations.

---

## Important Issues (Should Fix)

### 6. `overwrite or True` always evaluates to True
**File:** `src/diffing/methods/activation_difference_lens/causal_effect.py:485`
**Confidence: 95%**

```python
overwrite: bool = ... or True  # Always True regardless of config
```

Likely intended to read the config value and only default to True if not set.

---

### 7. Division by zero risk in token relevance
**File:** `src/diffing/methods/diff_mining/diff_mining.py:1250-1256`
**Confidence: 90%**

```python
total_w = sum(abs(w) for w in token_weights)
weighted_percentage = relevant_w / total_w if total_w > 0 else 0.0
```

Guard exists but `total_w == 0` indicates a degenerate NMF ordering (all weights filtered by the `< 1e-6` threshold at line 370). This should at minimum log a warning.

---

### 8. Missing shape assertions in vectorized functions
**File:** `src/diffing/methods/diff_mining/core_analysis.py:31-53+`
**Confidence: 85%**

Functions like `vectorized_bincount_masked`, `vectorized_shortlist_counts`, `vectorized_cooccurrence_shortlist`, and `vectorized_same_sign_cooccurrence` unpack tensor shapes without asserting dimensionality. Wrong-shaped input gives cryptic "too many values to unpack" errors. Per project guidelines, add assertions:
```python
assert indices.ndim == 3, f"indices must be 3D, got {indices.shape}"
```

---

### 9. JSONDecodeError retry logic may mask API issues
**File:** `src/diffing/utils/agents/llm.py:84-90`
**Confidence: 85%**

The `AgentLLM.chat()` retries on `json.JSONDecodeError` with exponential backoff, but the OpenAI client returns structured objects. If the error comes from internal JSON parsing of a malformed response, retrying the same request likely won't help. Should either remove or document when this legitimately occurs.

---

### 10. Hardcoded absolute path in steering script
**File:** `scripts/steering_from_logit_diff.py:262`
**Confidence: 95%**

```python
default="/mnt/nw/home/j.minder/.cache/huggingface/hub/models--auditing-agents--..."
```

User-specific path won't work for anyone else. Make the argument required or use a HuggingFace model ID instead.

---

### 11. API key setup not documented
**Files:** Multiple configs reference `api_key_path: openrouter_api_key.txt`
**Confidence: 85%**

Six config files reference this file but there's no documentation on how to create it. New users will hit cryptic errors.

---

## Minor Issues / Code Quality

### 12. Type annotation inconsistency: `float | None` vs `Optional[float]`
**Files:** `diff_mining.py:154-157`, `agent_tools.py:18`

PEP 604 syntax requires Python 3.10+. Rest of codebase uses `Optional[T]`.

### 13. Unused config assertions in `_ordering_dir_name`
**File:** `diff_mining.py:396-436`

Asserts that config sections exist but doesn't use their values (returns hardcoded strings). Either remove the assertions or use the config values.

### 14. `diff_mining.py` is 1752 lines
Could be split into `diff_mining_core.py`, `diff_mining_slr.py` (sequence likelihood ratio), and `diff_mining_relevance.py` (token relevance grading). Not blocking, but would improve navigability.

### 15. API key exposure risk in assertions
**Files:** `llm.py:45`, `grader.py:36`

`assert len(api_key) > 0` could leak key context in tracebacks. Use `ValueError` with a clean message instead.

### 16. DUPLICATION_REPORT.md may be stale
**File:** `docs/DUPLICATION_REPORT.md`

Generated 2025-02-05 with specific line numbers that have likely shifted. Should note the commit hash or be tracked as an issue.

---

## Architecture Observations (No Action Required)

### Positive

- **vllm.py deletion is safe** — no remaining imports, functionality consolidated into `DiffingMethod` base class
- **GPU memory management is correct** — auto-detection of utilization, halving for dual-server setups, proper cache clearing
- **LoRA adapter handling is well-implemented** — base server sharing, proper `LoRARequest` usage
- **diff_mining vectorization is excellent** — all batch ops use tensor operations, no Python loops over tokens
- **Agent budget tracking is robust** — multiple budgets (calls, tokens, model interactions) with clear assertions
- **New organism configs are consistent** — all follow established patterns
- **ADD_NEW_METHOD.MD is accurate** — cross-checked against actual implementation
- **All existing method changes are backward-compatible** — additive only, with proper defaults
- **get_agent() implementations are consistent** — either custom agent or `NotImplementedError`

### Design Notes

- `clear_nnsight_on_vllm_init` (default True) silently clears both nnsight models when either vLLM property is accessed. This is documented only in a code comment — consider adding to class docstring.
- Token relevance grading with `permutations=3` and retries can make up to 12 API calls per grading request. Cost should be documented.
- The experiment scripts have substantial code duplication (~4 scripts share `build_datasets_override()`, `run_command()`, etc.), but this is acceptable for self-contained research scripts.

---

## Files Not Reviewed

- `tests/` (excluded per request)
- `uv.lock` (auto-generated)

---

## Summary Table

| # | Severity | File | Issue | Confidence |
|---|----------|------|-------|------------|
| 1 | **Critical** | `base_agent.py:314` | `total_tokens` doubles instead of accumulating | 95% |
| 2 | **Critical** | `configs.py:142-168` | `disable_compile` not propagated to ModelConfig | 100% |
| 3 | **Critical** | `base_agent.py:426-440` | Silent TypeError catch hides tool bugs | 95% |
| 4 | **Critical** | `token_ordering.py:503-510` | `.todense()` on potentially huge sparse matrix | 95% |
| 5 | **Critical** | `CLAUDE.md:8` | Default method docs don't match config | 100% |
| 6 | Important | `causal_effect.py:485` | `overwrite or True` always True | 95% |
| 7 | Important | `diff_mining.py:1250` | Division by zero if NMF weights all zero | 90% |
| 8 | Important | `core_analysis.py:31+` | Missing tensor shape assertions | 85% |
| 9 | Important | `llm.py:84-90` | JSONDecodeError retry may not help | 85% |
| 10 | Important | `steering_from_logit_diff.py:262` | Hardcoded absolute user path | 95% |
| 11 | Important | Multiple configs | API key setup not documented | 85% |
| 12 | Minor | `diff_mining.py:154` | `float \| None` vs `Optional[float]` | 85% |
| 13 | Minor | `diff_mining.py:396` | Unused config assertions | 85% |
| 14 | Minor | `diff_mining.py` | 1752 lines, could be split | 80% |
| 15 | Minor | `llm.py:45` | API key in assertion traceback | 82% |
| 16 | Minor | `DUPLICATION_REPORT.md` | Line numbers may be stale | 75% |
